experimental/LieAlgebras/src/LieAlgebra.jl

```diff
@@ -432,6 +432,321 @@ Return `true` if `L` is solvable, i.e. the derived series of `L` terminates in $
+      end
+    end
+    B = B_new
+  end
+  filter!(!=(H), B)
+  @req all(B_j -> dim(B_j) == 1, B) "The Cartan subalgebra is not split"
+
+  # compute the roots, i.e. the list of eigenvalues of basis(H) on each B_j
+  root_spaces = Dict(
+    begin
+      b = only(basis(B_j))
+      root = [only(solve(_matrix(b), _matrix(bracket(h, b)); side=:left)) for h in basis(H)]
+      root => B_j
+    end for B_j in B
+  )
+  roots = collect(keys(root_spaces))
+
+  # compute an R-basis of the root space, s.t. the corresponding co-roots are a basis of H
+  roots_basis = empty(roots)
+  basis_mat_H = zero_matrix(F, 0, dim(L))
+  for root in roots
+    nrows(basis_mat_H) == dim(H) && break
+    x_j = only(basis(root_spaces[root]))
+    y_j = only(basis(root_spaces[-root]))
+    h_j = bracket(x_j, y_j)
+    if !can_solve(basis_mat_H, _matrix(h_j); side=:left)
+      basis_mat_H = vcat(basis_mat_H, _matrix(h_j))
+      push!(roots_basis, root)
+    end
+  end
+
+  function CartanInt(roots::Vector{RootType}, a::RootType, b::RootType) where {RootType}
+    # `a` and `b` are two roots in `roots`.
+    a == b && return 2
+    a == -b && return -2
+    # If a != ±b, the Cartan integer of `a` and `b` is `s-t`, where
+    # `s` and `t` are the largest integers such that `b-s*a` and `b+t*a` are still roots.
+    rt = b - a
+    s = 0
+    while rt in roots
+      s += 1
+      rt -= a
+    end
+    rt = b + a
+    t = 0
+    while rt in roots
+      t += 1
+      rt += a
+    end
+    return s - t
+  end
+
+  # we define a root to be positive if the first of its non-zero Cartan integers with the R-basis is positive
+  roots_positive = empty(roots)
+  for root in roots
+    2 * length(roots_positive) == length(roots) && break
+    root in roots_positive && continue
+    -root in roots_positive && continue
+    c = first(
+      Iterators.dropwhile(
+        iszero, Iterators.map(b_j -> CartanInt(roots, root, b_j), roots_basis)
+      ),
+    )
+    if c > 0
+      push!(roots_positive, root)
+    else
+      push!(roots_positive, -root)
+    end
+  end
+
+  # a positive root is simple if it is not the sum of two positive roots
+  roots_simple = empty(roots)
+  roots_positive_sums = Set(
+    alpha_i + alpha_j for (i, alpha_i) in enumerate(roots_positive) for
+    (j, alpha_j) in enumerate(roots_positive) if i < j
+  )
+  for root in roots_positive
+    if !(root in roots_positive_sums)
+      push!(roots_simple, root)
+    end
+  end
+  @assert length(roots_simple) == dim(H)
+
+  # compute the Cartan matrix and abstract root system
+  cm = matrix(
+    ZZ,
+    [
+      CartanInt(roots, alpha_i, alpha_j) for alpha_i in roots_simple,
+      alpha_j in roots_simple
+    ],
+  )
+  R = root_system(cm; check=true) # TODO: disable check
+
+  # reorder the simple roots according to the Dynkin diagram.
+  # users usually expect this ordering, but it is not necessary for the root system.
+  _, ordering = root_system_type_with_ordering(R)
```

Comment:
```suggestion
  _, ordering = cartan_type_with_ordering(cm)
```

This should also allow you to permute the matrix in place, we also don't compute the positive roots twice.

Reply:
You might also want to store the type and then set it manually after the root system was constructed.

Reply:
Good catch, thanks! I didn't think about this. Let's try it out